### PR TITLE
Improved check against integer overflow, including more borderline te…

### LINKED ISSUE #36: test_rdv_param_setting fails for some compilers
Test `test_rdv_param_setting.sli` fails when NEST is compiled with g++ 5.1.0 or Apple clang 6.1.0 with `-O2`.

The reason for this is unsafe integer-overflow detection in `librandom::UniformIntRandomDev::set_status()`, which leads to some overflow cases not being detected.

@apeyser provided a pointer to a safe implementation at https://www.securecoding.cert.org/confluence/display/c/INT32-C.+Ensure+that+operations+on+signed+integers+do+not+result+in+overflow 

Since there is a failing test (provided one uses the compilers given above, currently not in the NEST CI system), I am not adding an additional regression test.

### PR DESCRIPTION
**Pull request by heplesser**
This PR should fix #36 by adding safe overflow tests.
@apeyser @tammoippen

### CODE REVIEW

**Comment by tammoippen** [comment]
#39 Fixed the TravisCI .yaml. To rerun the TravisCI build on the most current source state, you have to do another commit in the heplesser:fix_36_test_rdv_param_setting_fails branch, e.g.

    git commit --amend   # this just touchs the last commit and does not modify anything
    git push origin fix_36_test_rdv_param_setting_fails

**Comment by tammoippen** [comment]
The code looks very nice to me. Thank you for the article on integer overflow. +1 for merge after TravisCI is green.

**Comment by abigailm** [comment]
cpp looks fine to me but I can't comment on the SLI, so a +1 with caveat.

File: testsuite/unittests/test_rdv_param_setting.sli [hunk 1]
```diff
@@ -244,20 +244,39 @@ statusdict /have_gsl :: {
   /MT19937 /uniform_int << /low 11 /high 10 >> run_test
 } fail_or_die
 
+/:range_limit
```
Comment (gewaltig) [inline testsuite/unittests/test_rdv_param_setting.sli]: I am not sure I follow what this function does. 
Can you maybe comment the parameter?

File: testsuite/unittests/test_rdv_param_setting.sli [hunk 2]
```diff
@@ -244,20 +244,39 @@ statusdict /have_gsl :: {
...
-  % of positive longs triggers an error; must be 32-bit safe
   % limit: 2^(N-1)-1, where N is number of bits for long
-  % build in overflow-safe way
-  /limit 
+  % compute in overflow-safe way
+
   statusdict/architecture/long :: 8 mul   % N
   2 sub 2 exch pow cvi                    % 2^(N-2)
   dup 1 sub % stack: 2^(N-2) 2^(N-2)-1
   add       % 2^(N-1)-1
```
Comment (gewaltig) [inline testsuite/unittests/test_rdv_param_setting.sli]: Calculation should be fine.

**Comment by heplesser** [comment]
@gewaltig I just pushed an updated version with a few more explanations.

**Comment by heplesser** [comment]
@gewaltig Could you check whether this PR is okay with the explanations I added on Monday and then give it a +1, so that the PR could be merged? Thank you!

**Comment by gewaltig** [comment]
Will do in 30 minutes.
Marc-Oliver

P.S. sent from a mobile device with small keyboard. Please excuse brevity and typos.

Am 24.06.2015 um 17:10 schrieb Hans Ekkehard Plesser <notifications@github.com<mailto:notifications@github.com>>:


@gewaltig<https://github.com/gewaltig> Could you check whether this PR is okay with the explanations I added on Monday and then give it a +1, so that the PR could be merged? Thank you!

File: testsuite/unittests/test_rdv_param_setting.sli (hunk 1 above)
Comment (gewaltig) [inline testsuite/unittests/test_rdv_param_setting.sli]: @heplesser My problem is the function :range_limit 
Can you add a comment, explaining what parameter it expects and what it is exactly computing in an overflow-safe way?

**Comment by heplesser** [comment]
@gewaltig I have now updated the code. The function always computed the same value, so instead of running the function, I just define a constant now. And I call that by its usual name for the value it stores, `:LONG_MAX`. I also added some more explanation of the way the value is computed.

**Comment by heplesser** [comment]
@gewaltig @lekshmideepu The Travis failure seems to be due to some technical trouble at Travis, all build variants failed after just 10 seconds or so. I will push a trivial change to trigger Travis once more.

**Comment by gewaltig** [comment]
SLI code looks good to me now. +1 for merging

**Pull request merged**